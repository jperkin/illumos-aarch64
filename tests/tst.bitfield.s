/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/*
 * C3.4.2 Bitfield
 */
.macro bitfield op
	/* Test for SBFM -> ASR/UBFM -> LSR (imms == 31/63) */
	\op	w0, wzr, #30, #31
	\op	x0, xzr, #31, #63
	/* Test for SBFM -> SBFIZ/BFM -> BFI (imms < immr) */
	\op	w0, wzr, #31, #0
	\op	x0, xzr, #63, #0
	/* Test for SBFM -> SXTB/UBFM -> UXTB (immr == 0 && imms == 7) */
	\op	w0, wzr, #0, #7
	\op	x0, xzr, #0, #7
	/* Test for SBFM -> SXTH/UBFM -> UXTH (immr == 0 && imms == 15) */
	\op	w0, wzr, #0, #15
	\op	x0, xzr, #0, #15
	/* Test for SBFM -> SXTW (immr == 0 && imms == 31) */
	\op	w0, wzr, #0, #31
	\op	x0, xzr, #0, #31
	/* Test for UBFM -> LSL (imms != 31/63 && immr == imms+1) */
	\op	w0, wzr, #31, #30
	\op	x0, xzr, #63, #62
	/* Test for SBFM -> SBFIZ (any operation not already matched above) */
	\op	w0, wzr, #30, #30
	\op	x0, xzr, #62, #62
.endm

.text
.globl libdis_test
.type libdis_test, @function
libdis_test:
	/*
	 * C5.6.158 SBFM
	 * C5.6.16 ASR (immediate) (alias)
	 * C5.6.157 SBFIZ (alias)
	 * C5.6.159 SBFX (alias)
	 * C5.6.201 SXTB (alias)
	 * C5.6.202 SXTH (alias)
	 * C5.6.203 SXTW (alias)
	 */
	bitfield sbfm
	/*
	 * C5.6.22 BFM
	 * C5.6.21 BFI (alias)
	 * C5.6.23 BFXIL (alias)
	 */
	bitfield bfm
	/*
	 * C5.6.212 UBFM
	 * C5.6.114 LSL (immediate) (alias)
	 * C5.6.117 LSR (immediate) (alias)
	 * C5.6.211 UBFIZ (alias)
	 * C5.6.213 UBFX (alias)
	 * C5.6.220 UXTB (alias)
	 * C5.6.221 UXTH (alias)
	 */
	bitfield ubfm
.size libdis_test, [.-libdis_test]
