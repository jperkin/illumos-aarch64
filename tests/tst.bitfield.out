    libdis_test:      e0 7f 1e 13  asr w0, wzr, #30
    libdis_test+0x4:  e0 ff 5f 93  asr x0, xzr, #31
    libdis_test+0x8:  e0 03 1f 13  sbfiz w0, wzr, #1, #1
    libdis_test+0xc:  e0 03 7f 93  sbfiz x0, xzr, #1, #1
    libdis_test+0x10: e0 1f 00 13  sxtb w0, wzr
    libdis_test+0x14: e0 1f 40 93  sxtb x0, wzr
    libdis_test+0x18: e0 3f 00 13  sxth w0, wzr
    libdis_test+0x1c: e0 3f 40 93  sxth x0, wzr
    libdis_test+0x20: e0 7f 00 13  asr w0, wzr, #0
    libdis_test+0x24: e0 7f 40 93  sxtw x0, wzr
    libdis_test+0x28: e0 7b 1f 13  sbfiz w0, wzr, #1, #31
    libdis_test+0x2c: e0 fb 7f 93  sbfiz x0, xzr, #1, #63
    libdis_test+0x30: e0 7b 1e 13  sbfx w0, wzr, #30, #1
    libdis_test+0x34: e0 fb 7e 93  sbfx x0, xzr, #62, #1
    libdis_test+0x38: e0 7f 1e 33  bfxil w0, wzr, #30, #2
    libdis_test+0x3c: e0 ff 5f b3  bfxil x0, xzr, #31, #33
    libdis_test+0x40: e0 03 1f 33  bfi w0, wzr, #1, #1
    libdis_test+0x44: e0 03 7f b3  bfi x0, xzr, #1, #1
    libdis_test+0x48: e0 1f 00 33  bfxil w0, wzr, #0, #8
    libdis_test+0x4c: e0 1f 40 b3  bfxil x0, xzr, #0, #8
    libdis_test+0x50: e0 3f 00 33  bfxil w0, wzr, #0, #16
    libdis_test+0x54: e0 3f 40 b3  bfxil x0, xzr, #0, #16
    libdis_test+0x58: e0 7f 00 33  bfxil w0, wzr, #0, #32
    libdis_test+0x5c: e0 7f 40 b3  bfxil x0, xzr, #0, #32
    libdis_test+0x60: e0 7b 1f 33  bfi w0, wzr, #1, #31
    libdis_test+0x64: e0 fb 7f b3  bfi x0, xzr, #1, #63
    libdis_test+0x68: e0 7b 1e 33  bfxil w0, wzr, #30, #1
    libdis_test+0x6c: e0 fb 7e b3  bfxil x0, xzr, #62, #1
    libdis_test+0x70: e0 7f 1e 53  lsr w0, wzr, #30
    libdis_test+0x74: e0 ff 5f d3  lsr x0, xzr, #31
    libdis_test+0x78: e0 03 1f 53  ubfiz w0, wzr, #1, #1
    libdis_test+0x7c: e0 03 7f d3  ubfiz x0, xzr, #1, #1
    libdis_test+0x80: e0 1f 00 53  uxtb w0, wzr
    libdis_test+0x84: e0 1f 40 d3  ubfx x0, xzr, #0, #8
    libdis_test+0x88: e0 3f 00 53  uxth w0, wzr
    libdis_test+0x8c: e0 3f 40 d3  ubfx x0, xzr, #0, #16
    libdis_test+0x90: e0 7f 00 53  lsr w0, wzr, #0
    libdis_test+0x94: e0 7f 40 d3  ubfx x0, xzr, #0, #32
    libdis_test+0x98: e0 7b 1f 53  lsl w0, wzr, #1
    libdis_test+0x9c: e0 fb 7f d3  lsl x0, xzr, #1
    libdis_test+0xa0: e0 7b 1e 53  ubfx w0, wzr, #30, #1
    libdis_test+0xa4: e0 fb 7e d3  ubfx x0, xzr, #62, #1
